Implement token bucket duty cycle enforcement

ViezeVingertjes opened the pull request:
Replaces per-transmission delay with a rolling window token bucket for better regulatory compliance.

Resolves #817.

ViezeVingertjes commented:
We had a discussion in the Dutch community today. Some people didn't care much, others wanted a proper implementation. I'm neutral either way (not even sure if I enabled it myself). Perhaps this could start a broader discussion to decide it once and for all.

LitBomb commented:
how does this work for USA where there is no duty cycle limit?

LitBomb commented:
whatever the change is, it should not limit duty cycle at any percentage for regions that don't have duty cycle limit.

ViezeVingertjes commented:
> > Hmm, but default AF was 1.0 right? So i feel 10% would reflect the original 'duty cycle' more than 0%? But disabled would be fine with me too though: [eec5ef0](https://github.com/meshcore-dev/MeshCore/commit/eec5ef085364770676496081fa2c478b0d36ccd2)
> 
> yes, you are right. AF=0 can be dangerous.. AF=1 is optimal, sorry for that.

Haha check! thought maybe my thought train was wrong, reverted it -> 1b959a913924136aabe528bb58ecd60594a1607e

ViezeVingertjes commented:
> > Haha check! thought maybe my thought train was wrong, reverted it -> [1b959a9](https://github.com/meshcore-dev/MeshCore/commit/1b959a913924136aabe528bb58ecd60594a1607e)
> 
> you reverted it to 9 tho? we need it at 1

Yes; i went off from numbers i saw on my repeaters which maxed out on 9%, but there are more things at play on there. Updated it to 1.0 and cleaned up the commits. a43b0e5aab823f0b8c15e86cbaac40d5d583cb38

fschrempf commented:
@ViezeVingertjes You can simply drop the last two commits, right? First you change it from 1 to 9 and then back from 9 to 1. Except for the `simple_secure_chat` which was at 2 for some reason at might want to be aligned with the others.

ViezeVingertjes committed:
Implement token bucket duty cycle enforcement

ViezeVingertjes commented:
> @ViezeVingertjes You can simply drop the last two commits, right? First you change it from 1 to 9 and then back from 9 to 1. Except for the `simple_secure_chat` which was at 2 for some reason at might want to be aligned with the others.

Done.

marcelverdult commented:
thanks so much for implementing my request!

marcelverdult commented:
@liamcottle @andymux this would be quiete important for countries with strict dutycycle limits. the current implementation with the waits throws away valuable tx time.

jbrazio commented:
@ViezeVingertjes just to make sure, the % calculation is still the same as the original one ?
I.e. AF1 = 50%, AF9 = 10% ?

weebl2000 commented on src/Dispatcher.cpp:
```diff
@@ -20,12 +20,34 @@ void Dispatcher::begin() {
+    tx_budget_ms = max_budget;
+  }
+
+  last_budget_update = now;
```
When `updateTxBudget()` is called on fast loop iterations (1-5ms elapsed) and `duty_cycle` is small (e.g. 0.1 for factor=9), `(unsigned long)(elapsed * duty_cycle)` truncates to 0 every iteration. Since `last_budget_update` is always advanced to `now`, the lost fractional milliseconds never accumulate — the budget effectively stops refilling.

Fix: only advance `last_budget_update` when `refill > 0`, so elapsed time accumulates across iterations until it produces a non-zero refill. Max error at any point is <1ms of budget.

```suggestion
  unsigned long refill = (unsigned long)(elapsed * duty_cycle);
  if (refill > 0) {
    tx_budget_ms += refill;
    if (tx_budget_ms > max_budget) {
      tx_budget_ms = max_budget;
    }
    last_budget_update = now;
  }
```

weebl2000 reviewed (commented):
Budget refill truncation fix

weebl2000 reviewed (commented):
Budget refill truncation fix

ViezeVingertjes committed:
Update logic in Dispatcher to ensure refill is only applied when greater than zero.

weebl2000 reviewed (approved):
✔️ Looks good. Maybe clarify two magic numbers with a small comment.

I think this is a big improvement for bursty workloads like room servers.

ViezeVingertjes committed:
Updated the Dispatcher logic to replace hardcoded values with defined constants for minimum TX budget reserve and airtime division.

ViezeVingertjes commented on src/Dispatcher.cpp:
```diff
@@ -61,11 +83,24 @@ void Dispatcher::loop() {
+        tx_budget_ms -= t;
+      }
+
+      if (tx_budget_ms < 100) {
```
https://github.com/meshcore-dev/MeshCore/pull/1297/changes/519b97a90ab9786e65423ac0d9dd7a9fd92dd180

ViezeVingertjes commented on src/Dispatcher.cpp:
```diff
@@ -224,9 +259,20 @@ void Dispatcher::processRecvPacket(Packet* pkt) {
+  updateTxBudget();
+  
+  uint32_t est_airtime = _radio->getEstAirtimeFor(MAX_TRANS_UNIT);
+  if (tx_budget_ms < est_airtime / 2) {
```
Half isnt too bad; but perhaps this helps -> https://github.com/meshcore-dev/MeshCore/pull/1297/changes/519b97a90ab9786e65423ac0d9dd7a9fd92dd180

weebl2000 commented:
My 2 cents would be to merge it into v1.14.0😋. I'm really confident it's only going to improve the mesh as a whole.

ripplebiz commented:
I am _for_ this change, in principle, but I'm a little skeptical of the implementation, particularly all the arithmetic with the millis(). So, I did an experiment and modified the ArduinoHelpers.h, starting the millis clock at 5 minutes before the 49-day/32-bit wrap-around event:

class ArduinoMillis : public mesh::MillisecondClock {
public:
  unsigned long getMillis() override { return 4294667296 + millis(); }  // TEMP!!   5 minutes before 49-day wrap around
};

And, for five minutes I couldn't login to my test repeater, but as soon as 5 minutes passed it all worked again. 
Please look into this, and how to work _with_ the 49-day wrap-around.
